File: src/algorithms/python/Palindrome_Linked_List.py
# -*- coding: utf-8 -*-


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # @param {ListNode} head
    # @return {boolean}
    def isPalindrome(self, head):
        if not head:
            return True
        node, length, end = head, 0, None
        while node:
            if not node.next:
                end = node
            node, length = node.next, length + 1
        node = head
        for i in range(length / 2):
            node = node.next
        pre = node
        node = node.next
        while node:
            _next = node.next
            node.next, pre = pre, node
            node = _next
        node1, node2 = head, end
        for i in range(length / 2):
            if node1.val != node2.val:
                return False
            node1, node2 = node1.next, node2.next
        return True


if __name__ == "__main__":
    solution = Solution()
    l = [1, 2, 3, 3, 1]
    l = map(ListNode, l)
    for i in range(len(l) - 1):
        l[i].next = l[i + 1]
    print solution.isPalindrome(l[0])

File: src/algorithms/cpp/Palindrome_Linked_List.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
private:
    ListNode *rec_node;
    bool result;

private:
    void search(ListNode *node) {
        if (node->next)     search(node->next);
        if (rec_node->val != node->val)     result = false;
        rec_node = rec_node->next;
    }

public:
    bool isPalindrome(ListNode* head) {
        result = true;
        rec_node = head;
        search(head);
        return result;
    }
};

int main() {
    return 0;
}
